[new-rule] prefer-user-event (#192)
testing-library/eslint-plugin-testing-library: docs/rules/prefer-user-event.md

# Use [userEvent](https://github.com/testing-library/user-event) over using `fireEvent` for user interactions (prefer-user-event)

From
[testing-library/dom-testing-library#107](https://github.com/testing-library/dom-testing-library/issues/107):

> [...] it is becoming apparent the need to express user actions on a web page
> using a higher-level abstraction than `fireEvent`

`userEvent` adds related event calls from browsers to make tests more realistic than its counterpart `fireEvent`, which is a low-level api.  

> See the appendix at the end to check how are the events from `fireEvent` mapped to `userEvent`

Reviewer: ```suggestion
See the appendix at the end to check how are the events from `fireEvent` mapped to `userEvent`.
```

## Rule Details

> This rules enforces the usage of [userEvent](https://github.com/testing-library/user-event) methods over `fireEvent`. By default, the methods from `userEvent` takes precedence, but you add exceptions by configuring the rule in `.eslintrc`

Reviewer: ```suggestion
This rule enforces the usage of [userEvent](https://github.com/testing-library/user-event) methods over `fireEvent`. By default, the methods from `userEvent` take precedence, but you add exceptions by configuring the rule in `.eslintrc`.
```

> See below for examples of valid usages of `fireEvent` methods with the configuration.

Reviewer: Personally, I think this line can be removed - or that it can be replaces with a link to the options.
I'm fine either way tho, this was the only comment I would not have said it.
```suggestion
```

Examples of **incorrect** code for this rule:

```ts
// a method in fireEvent that has a userEvent equivalent
import { fireEvent } from '@testing-library/dom';
fireEvent.click(node);

// using fireEvent with an alias
import { fireEvent as fireEventAliased } from '@testing-library/dom';
fireEventAliased.click(node);

// using fireEvent after importing the entire library
import * as dom from '@testing-library/dom';
dom.fireEvent.click(node);
```

Examples of **correct** code for this rule:

```ts

> // any userEvent method

Reviewer: Because most samples include an import, I would add it here too:

```suggestion
import userEvent from '@testing-library/user-event';
// any userEvent method
```

userEvent.click();

> // fireEvent method that does not have an alternative in userEvent

Reviewer: Sorry, I feel like a jerk to nit about the style, but since this is public I think it should look right 😅

```suggestion

// fireEvent method that does not have an alternative in userEvent
```

fireEvent.cut(node);

> import * as dom from '@testing-library/dom';

Reviewer: ```suggestion

import * as dom from '@testing-library/dom';
```

dom.fireEvent.cut(node);

> // a function called fireEvent that's not imported from testing-library
> function fireEvent() {
>   // do stuff
> }
> fireEvent();

Reviewer: I think we can assume this is the case for most of the rules, thus I would remove it.
It sure is important for us (in our tests), but I don't see an added value to add it to the docs.

```suggestion
```

Author: iirc it was just for the sake of coverage, otherwise, I wouldn't get the 100%... but I changed so many times the tests that I'll review it again to see if I can remove it.

Reviewer: It sure can exists in the specs.
But it shouldn't live in the docs, imho.

Author: oh lol, I just reviewed the comment from the emails and I thought you were commenting in the test, not in the docs. Yeah, I think that makes sense. I'll remove it

This rule allows to exclude specific functions with an equivalent in `userEvent` through configuration. This is useful if you need to allow an event from `fireEvent` to be used in the solution. For specific scenarios, you might want to consider disabling the rule inline.

The configuration consists of an array of strings with the names of fireEvents methods to be excluded.  
An example looks like this

```json
{
  "rules": {
    "prefer-user-event": [
      "error",
      {
        "allowedMethods": ["click", "change"]
      }
    ]
  }
}
```

With this configuration example, the following use cases are considered valid

```ts
// using a named import
import { fireEvent } from '@testing-library/dom';
fireEvent.click(node);
fireEvent.change(node, { target: { value: 'foo' } });

// using fireEvent with an alias
import { fireEvent as fireEventAliased } from '@testing-library/dom';
fireEventAliased.click(node);
fireEventAliased.change(node, { target: { value: 'foo' } });

// using fireEvent after importing the entire library
import * as dom from '@testing-library/dom';
dom.fireEvent.click(node);
dom.fireEvent.change(node, { target: { value: 'foo' } });
```

## When Not To Use It

> When you don't want to encourage the usage of `userEvent`

Author: I did not know what to put here 😅  🤷

Reviewer 2: Maybe simplify this a little

```suggestion
When you don't want to use `userEvent`, such as if a legacy codebase is still using `fireEvent` or you need to have more low-level control over firing events (rather than the recommended approach of testing from a user's perspective)
```

| `fireEvent.pointerEnter()` | <ul><li>`userEvent.hover()`</li><li>`userEvent.selectOptions()`</li><li>`userEvent.deselectOptions()`</li></ul>                                                         |
| `fireEvent.pointerLeave()` | <ul><li>`userEvent.unhover()`</li></ul>                                                                                                                                 |
| `fireEvent.pointerMove()`  | <ul><li>`userEvent.hover()`</li><li>`userEvent.unhover()`</li><li>`userEvent.selectOptions()`</li><li>`userEvent.deselectOptions()`</li></ul>                           |
| `fireEvent.pointerOut()`   | <ul><li>`userEvent.unhover()`</li></ul>                                                                                                                                 |
| `fireEvent.pointerOver()`  | <ul><li>`userEvent.hover()`</li><li>`userEvent.selectOptions()`</li><li>`userEvent.deselectOptions()`</li></ul>                                                         |

> | `fireEvent.pointerUp()`    | <ul><li>`userEvent.click()`</li><li>`userEvent.dblClick()`</li><li>`userEvent.selectOptions()`</li><li>`userEvent.deselectOptions()`</li></ul>                          |

Reviewer 2: I don't think we need to repeat `fireEvent`/`userEvent` and parentheses in this table.
